gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+		}
+		if !osutil.FileExists(backupName + ".backup") {
+			// not preserver & different than the update, double
+			// check we have a backup
+			return fmt.Errorf("missing backup file for %v", dstName)
+		}
+	}
+
+	return writeFile(srcName, dstName, preserveInDst)
+}
+
+func (f *MountedFilesystemUpdater) updateVolumeContent(whereDir string, content *VolumeContent, preserveInDst []string, backupDir string) error {
+	if err := checkContent(content); err != nil {
+		return err
+	}
+
+	srcName, _, _ := f.entryNames(whereDir, content.Source, content.Target, backupDir)
+
+	if osutil.IsDirectory(srcName) || strings.HasSuffix(content.Source, "/") {
+		return f.updateDirectory(whereDir, content.Source, content.Target, preserveInDst, backupDir)
+	} else {
+		return f.updateOrSkipFile(whereDir, content.Source, content.Target, preserveInDst, backupDir)
+	}
+}
+
+// Backup analyzes a mounted filesystem and prepares a rollback state
+// information should the update be applied. The content of the filesystem is
+// processed, files that would be modified by the update are backed up, while
+// other files may be stamped to improve the later step of update process.
+func (f *MountedFilesystemUpdater) Backup() error {
+	mount, err := f.mountLookup(f.ps)
+	if err != nil {
+		return fmt.Errorf("cannot find mount location of structure %v: %v", f.ps, err)
+	}
+
+	backupRoot := fsStructBackupPath(f.backupDir, f.ps)
+
+	if err := os.MkdirAll(backupRoot, 0755); err != nil {
+		return fmt.Errorf("cannot create backup directory: %v", err)
+	}
+
+	for _, c := range f.ps.Content {
+		if err := f.backupOrCheckpointVolumeContent(mount, &c, prefixPreserve(mount, f.ps.Update.Preserve), backupRoot); err != nil {
+			return fmt.Errorf("cannot backup content: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointDirectory(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcName := f.entrySource(source)
+
+	if err := checkSrcIsDir(srcName); err != nil {
+		return fmt.Errorf("cannot backup directory %q: %v", source, err)
+	}
+
+	fis, err := ioutil.ReadDir(srcName)
+	if err != nil {
+		return fmt.Errorf("cannot list directory entries: %v", err)
+	}
+
+	if !strings.HasSuffix(source, "/") {
+		// /foo -> /bar/ => /bar/foo
+		target = filepath.Join(target, filepath.Base(source))
+	}
+
+	for _, fi := range fis {
+		pSrc := filepath.Join(source, fi.Name())
+		pDst := filepath.Join(target, fi.Name())
+
+		backup := f.backupOrCheckpointFile
+		if fi.IsDir() {
+			if err := f.backupOrCheckpointPrefix(dstRoot, pDst+"/", backupDir); err != nil {
+				return fmt.Errorf("cannot backup prefix directory: %v", err)
+			}
+
+			backup = f.backupOrCheckpointDirectory
+			pSrc += "/"
+		}
+		if err := backup(dstRoot, pSrc, pDst, preserveInDst, backupDir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) backupOrCheckpointPrefix(dstRoot, target string, backupDir string) error {
```

Comment:
this probably merits a comment about what it  really does, is just making a stamp afaict ? is that true?

Reply:
Added a comment. Tweaked naming and the code a little to make it easier to follow. Also call sites make it clearer now.